### ISSUE TITLE
Process invitations via a Lambda hook to automate / speed up invites

### ISSUE THREAD

**Issue by kaihendry** [unee-t/bugzilla-customisation#46]
Before sending out the invitation, a process on Bugzilla needs to be performed which I am **unclear about**.

Once it's done, I know I send the invitation off with `http://localhost:3000/api/pending-invitations/done?accessToken=blablabla`...

**Comment by franck-boullier** [comment]
yep, I need to do that: there are a few scripts in the bz-database repo that we need to run.
There are a lot of comment inside the scripts which should be a good start to understand what need to happen. 
I need to create an overview and connect all the dots though, this would be needed when we automate the invitation process (hopefully soon...)

**Comment by kaihendry** [comment]
Any updates on this? This is quite important for testing the new signup flow.

**Comment by franck-boullier** [comment]
Will work on that over the WE, sorry for the delay

**Comment by franck-boullier** [comment]
@kaihendry 
Let me know if you have questions on [this explanation](https://github.com/unee-t/bugzilla-customisation/wiki/Process-Invitations-in-the-BZ-database).

**Comment by kaihendry** [comment]
So next step is to implement https://github.com/unee-t/bugzilla-customisation/wiki/Process-Invitations-in-the-BZ-database as an automated function right? I could do that.

**Comment by franck-boullier** [comment]
Yes
Next step is to automate that and that'll be great if you can do that!

**Comment by kaihendry** [comment]
Made a video about how I setup my dev environment for this: https://s.natalian.org/2018-04-10/setup-demo.mp4

**Comment by kaihendry** [comment]
Created a new repo to address this issue. https://github.com/unee-t/processInvitations

It's a Apex Up style lambda function.

**Comment by franck-boullier** [comment]
The database schema version that will allow this to work will be v3.0
https://github.com/unee-t/bz-database/milestone/2

**Comment by franck-boullier** [comment]
>Created a new repo to address this issue. https://github.com/unee-t/processInvitations 

@kaihendry can you setup this repo so that it post to Slack when updated too?

**Comment by kaihendry** [comment]
Slack integration is done. Are https://github.com/unee-t/bz-database/commit/76274ec91bdf56e1b6465fded5a1da94e185129e the scripts I need to [exec](https://github.com/unee-t/processInvitations/blob/78c3a63d70d7dda8e3c5cfba414db20bb4e2e22b/main.go#L201) after the [INSERT](https://github.com/unee-t/processInvitations/blob/78c3a63d70d7dda8e3c5cfba414db20bb4e2e22b/main.go#L84)?

**Comment by franck-boullier** [comment]
not yet, I need to test them first...
I'll do that now
Once my tests are passed, I'll merge into the master and point you to the script you need to run

**Comment by franck-boullier** [comment]
also these scripts will only work on a BZ DB which run the v3.0 of the Unee-T BZ DB schema so I also need to  test and do that update too first.

**Comment by franck-boullier** [comment]
Waiting for https://github.com/unee-t/bz-database/issues/30 to be done so we can update the Schema of the BZ database to v3.0

**Comment by kaihendry** [comment]
But we have Aurora running in staging, so we shouldn't be blocked by a Prod DB migration, should we?

**Comment by franck-boullier** [comment]
> But we have Aurora running in staging, so we shouldn't be blocked by a Prod DB migration, should we?

To do the tests in the DEV/Staging of course not
To move all this in prod once all the tests are done we need to make sure that both environments (DEV/Staging and prod) are similar. This is why I feel this is important: the ultimate goal is to do that in prod...

**Comment by franck-boullier** [comment]
# Summary of what needs to happen:
1- A user creates an invitation in the MEFE

2- A record with all the information about the invitation is inserted in the BZ DB table 'ut_invitation_api_data' 

3- The script 'process_one_invitation_all_scenario_v3.0.sql' (https://github.com/unee-t/bz-database/blob/master/scripts/process_one_invitation_all_scenario_v3.0.sql) is triggered. This script needs 2 variables:
- @mefe_invitation_id (line 10 of the script)
- @environment (line 16 of the script)

4- Once the script has ran with no error MEFE is notified that the invitation has been processed in the BZ database and sends the email notification to the invitee.

5- Once the invitation has been sent to the invitee by the MEFE, a script 'MEFE_invitation_sent_v3.0.sql' (NOT created yet) is triggered. This script needs 2 variables:
- @mefe_invitation_id
- @environment.

This script will:
- Update the table 'ut_invitation_api_data' 
- insert a new message in the case to let the users know that the invitation has been sent to the invited user.

6- Move to the next invitation to process and start over from step 2.

# Error handling:
At this point we need to have a notification sent to a human if there was an error during this procedure. The message should mention:
- MEFE invitation_id
- Datetime when the invitation was created in the MEFE by the user
- Datetime of the error
- Which step failed in this process.

# Status:
In the DEV/Staging for the BZ database, all the objects needed until step 4 are created. 
## The missing parts:
For the BZ database:
- the script 'MEFE_invitation_sent_v3.0.sql' to do step 5 above.
- error handling to roll back step 2 if there is an error during step 3.
